apache/superset
# feat(advanced analysis): support MultiIndex column in post processing stage

### Summary
Superset uses Pandas Dataframe for query results processing. Now, these Dataframe between different [Operators](https://github.com/apache/superset/tree/master/superset/utils/pandas_postprocessing) only support 1-dimensional Dataframe. This has caused confusion in some calculations. This PR introduces MultiIndex Dataframe for Operators. This is useful for simplifying calculations.

### Current PostProcessing
Typical PostProcessing Operators are used in [time-series charts](https://github.com/apache/superset/blob/e3e03d202db959d708acbfe8a25a50dddc85fa7f/superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/buildQuery.ts#L73) for getting queries. e.g.:
```
        post_processing: [
          resampleOperator(formData, baseQueryObject),
          timeCompareOperator(formData, baseQueryObject),
          sortOperator(formData, { ...baseQueryObject, is_timeseries: true }),
          pivotOperatorInRuntime,
          rollingWindowOperator(formData, baseQueryObject),
          contributionOperator(formData, baseQueryObject),
          prophetOperator(formData, baseQueryObject),
        ],
```
We face these challenges for existing designs: 
1. Such query can't be adapted to both rolling calculation and time compared calculation. In the other words, `rollingWindowOperator` __should__ before `timeCompareOperator`. 
2. We also need to consider how to face series in Dataframe. [for example](https://github.com/apache/superset/blob/e3e03d202db959d708acbfe8a25a50dddc85fa7f/superset/utils/pandas_postprocessing/resample.py#L30), in the `resample.py`. We need to "flat" multidimensional Dataframe to adapt this one.


### After Changed
The new operators support calculate on multidimensional Dataframe. So, operators in `QueryObject` can change its position at will. eg:
```
        post_processing: [
          pivotOperatorInRuntime,
          resampleOperator(formData, baseQueryObject),
          rollingWindowOperator(formData, baseQueryObject),
          timeCompareOperator(formData, baseQueryObject),
          flatOperator(formData, baseQueryObject),
          contributionOperator(formData, baseQueryObject),
          prophetOperator(formData, baseQueryObject),
        ],
```

For example, The `compare()` [operator](https://github.com/apache/superset/blob/818eb02fc045100b02208173f4d889e309fd323d/superset/utils/pandas_postprocessing/compare.py) supports the Dataframe shape like the following:

```
                   count_metric    sum_metric
    country              UK US         UK US
    dttm
    2019-01-01            1  2          5  6
    2019-01-02            3  4          7  8
```

adding new operator `flat.py`, this new operator will transform the multidimensional dataframe to flatten dataframe.

The full examples:
```
    pivot_df = pp.pivot(
        df=multiple_metrics_df,
        index=["dttm"],
        columns=["country"],
        aggregates={
            "sum_metric": {"operator": "sum"},
            "count_metric": {"operator": "sum"},
        },
        flatten_columns=False,
        reset_index=False,
    )
    """
                   count_metric    sum_metric
    country              UK US         UK US
    dttm
    2019-01-01            1  2          5  6
    2019-01-02            3  4          7  8
    """
    compared_df = pp.compare(
        pivot_df,
        source_columns=["count_metric"],
        compare_columns=["sum_metric"],
        compare_type=PPC.DIFF,
        drop_original_columns=True,
    )
    """
               difference__count_metric__sum_metric
    country                                      UK US
    dttm
    2019-01-01                                    4  4
    2019-01-02                                    4  4
    """
    flat_df = pp.flat(compared_df)
    """
            dttm  difference__count_metric__sum_metric, UK  difference__count_metric__sum_metric, US
    0 2019-01-01                                         4                                         4
    1 2019-01-02                                         4                                         4
    """
```

### TESTING INSTRUCTIONS
Let's use a new [dataset](https://www.kaggle.com/sumanthvrao/daily-climate-time-series-data?select=DailyDelhiClimateTest.csv) to test.
1. download `DailyDelhiClimateTrain.csv` to your local.
2. import `DailyDelhiClimateTrain` into Superset
  a) click `upload CSV to database` on top blue plus sign of Superset
  b) fill in `DailyDelhiClimateTrain` on `table name`
  c) select `DailyDelhiClimateTrain.csv`
  d) fill in `date` in `Parse Dates` line
  e) click Save
3. open explore page and select `DailyDelhiClimateTrain` dataset
4. change viz type to `Line Chart`, __notice that__ you should pick up echart version
5. drag `date` in `X-Axis`
6. drag `meantemp` in `metrics` and select `max` for aggregate
7. select a time range
![image](https://user-images.githubusercontent.com/2016594/158374453-4e5e70d7-c067-4915-81db-5541194d3415.png)
8. select `day` as time grainularity
9. select a time shift in AA section `1 year ago` and `actual values`
10. validate data in 2015 and data in 2014 are correct (use SQLLab to validate)
11. select different calculation types in time comparison and validate data
12. select different rolling types in `rolling window` and validate data
13. select different time granularity and validate data.
14. you can find some "time hole" and validate Resample

Daily PCT with cumsum
![image](https://user-images.githubusercontent.com/2016594/158376224-c465c758-de0e-448e-a48a-6f2fb7906cb1.png)

Monthly comparison with quarter sum rolling
![image](https://user-images.githubusercontent.com/2016594/158376467-6e747e1f-da5b-49ae-9a50-da2a613ff153.png)

### ADDITIONAL INFORMATION
